examples/autoignition-mpi.py

```diff
@@ -244,67 +298,144 @@ def main(ctx_factory=cl.create_some_context, casename="autoignition", use_leap=F
-        if rank == 0:
-            logger.info("Checkpointing final state ...")
-        my_checkpoint(current_step, t=current_t,
-                      dt=(current_t - checkpoint_t),
-                      state=current_state)
-
-    if current_t - t_final < 0:
-        error_state = True
-
-    if error_state:
-        raise ValueError("Simulation did not complete successfully.")
+    def my_post_step(step, t, dt, state):
+        # Logmgr needs to know about EOS, dt, dim?
+        # imo this is a design/scope flaw
+        if logmgr:
+            set_dt(logmgr, dt)
+            set_sim_state(logmgr, dim, state, eos)
+            logmgr.tick_after()
+        return state, dt
+
+    def my_pre_step(step, t, dt, state):
+        dv = None
+        pre_step_errors = False
+
+        if logmgr:
+            logmgr.tick_before()
+
+        from mirgecom.simutil import check_step
+        do_viz = check_step(step=step, interval=nviz)
+        do_restart = check_step(step=step, interval=nrestart)
+        do_health = check_step(step=step, interval=nhealth)
+
+        if step == rst_step:  # don't do viz or restart @ restart
+            do_viz = False
+            do_restart = False
+
+        if do_health:
+            dv = eos.dependent_vars(state)
+            local_health_error = my_health_check(dv, dt)
+            health_errors = False
+            if comm is not None:
+                health_errors = comm.allreduce(local_health_error, op=MPI.LOR)
+            if health_errors and rank == 0:
+                logger.info("Fluid solution failed health check.")
+            pre_step_errors = pre_step_errors or health_errors
+
+        if do_restart:
+            my_write_restart(state, step, t)
+
+        if do_viz:
+            production_rates = eos.get_production_rates(state)
+            if dv is None:
+                dv = eos.dependent_vars(state)
+            my_write_viz(cv=state, dv=dv, step=step, t=t,
+                         production_rates=production_rates)
+
+        if pre_step_errors:
+            my_graceful_exit(cv=state, step=step, t=t,
+                             do_viz=(not do_viz), do_restart=(not do_restart),
+                             message="Error detected at prestep, exiting.")
+
+        return state, dt
+
+    current_step, current_t, current_state = \
+        advance_state(rhs=my_rhs, timestepper=timestepper,
+                      pre_step_callback=my_pre_step,
+                      post_step_callback=my_post_step,
+                      get_timestep=get_timestep, state=current_state,
+                      t=current_t, t_final=t_final, eos=eos, dim=dim)
+
+    finish_tol = 1e-16
+    if np.abs(current_t - t_final) > finish_tol:
+        my_graceful_exit(cv=current_state, step=current_step, t=current_t,
+                         do_viz=True, do_restart=True,
+                         message="Simulation timestepping did not complete.")
```

Comment:
I don't think.

Reply:
I don't see a case where the stepper would just exit prematurely... either it's going to finish, or it's going to raise an exception.

Reply:
This has already caught some issues for me, where the stepper stepped significantly beyond t_final.  The stepper will return early if dt is negative.

Reply:
Looks like the negative `dt` check is gone now, right?

If the stepper advanced beyond where it should, that seems like a bug... maybe this check could be in a test? I'm not sure it needs to be in every driver.

Reply:
It was a indeed a bug, and this check caught it, which is why I'm a little bit dubious of removing it.  Since it is doing no harm, can we keep it for just a bit?

Reply:
Ehhh... how about just leaving a:

```diff
-    finish_tol = 1e-16
-    if np.abs(current_t - t_final) > finish_tol:
-        my_graceful_exit(cv=current_state, step=current_step, t=current_t,
-                         do_viz=True, do_restart=True,
-                         message="Simulation timestepping did not complete.")
+    assert np.abs(current_t - t_final) <= 1e-16
```

as a sanity check?

Reply:
I like that, but I don't get a final dump then.

Reply:
I think if it moved to after the final dump, then it is superior to what I had.

Reply:
How's this? (ea98b59)